### ISSUE TITLE
Latest RStudio Update Uses readr As Default Point-And-Click Importer

### ISSUE THREAD

**Issue by Geeketics** [swcarpentry/r-novice-inflammation#276]
In the latest RStudio update the readr package has been linked to the "Import Dataset" button in RStudio, this used to be the utils package. This can cause problems downstream, particularly with subsetting.

I know SWC lessons are all about making people learn to code things, instead of point-and-click, but it seems important to add a note to all the R lessons that warns the learner that the point-and-click "Import Dataset" button in RStudio may not do what they expect it to and that this could cause problems later on. Basically it would be emphasising the need to actually type in the code, rather than trying to take a point-and-click short cut.

**Comment by chendaniely** [comment]
thanks for the heads up. do you know where we should add this comment?

I feel like it needs to be mentioned multiple times in the lesson:

1. while importing data
2. when we talk about `as.data.frame`

I don't know off the top of my head what the differences between the `tibble` object and the regular `data.frame` object is...

**Comment by bbolker** [comment]
The main/most import difference I'm aware of is that tibbles do *not* drop dimensions by default as other R rectangular data structures do: in particular, `tbl[,"foo"]` is a tibble rather than a vector (one should use `tbl[["foo"]]` to get a vector). That being said, it's probably sufficient to tell beginning students "tibbles usually but **not always** act the same as data frames: if necessary, use `as.data.frame()` ..."

**Comment by katrinleinweber** [comment]
Wouldn't this instance even be a starting point to start converting the lesson to tibbles? Continuing to teach a non-default (compared to the most popular R-IDE) might cause more problems than it's worth.

**Comment by pauljohn32** [comment]
I'm not an RStudio user/enthusiast.  I think it imposes the particular opinions and packages of the authors onto the users in a restrictive way.  Given that, you may not be surprised with my counter argument: if the software imposes judgments that make you feel you need to teach about different code content,  you should consider getting a different software tool.  You would not put up with a C++ editor that imposes somebody's library of routines for floating point calculations on you. Would you?  In that case, I'd say use a different editor.  R existed long before RStudio and many of us who write R packages don't use RStudio at all.

**Comment by bbolker** [comment]
I'm of two minds. On the one hand, the creeping tidyverse annoys me. On the other hand, I have to admit that some of `read_csv`'s defaults are nice (strings not automatically converted to factors; dates automatically converted to dates; flexible type-specification; informative report about parsing errors). Bugs me that there's not an argument to specify that column names should be sanitized (https://github.com/tidyverse/readr/issues/388).  

In short, I would suggest that the scope of this issue be limited to @Geeketics's original point: we should warn learners (briefly/unobtrusively) that point-and-click import will give different results from `read.csv()`. (If the lessons start to use `readr::read_csv()`, then alternatively we need to let learners know that these give subtly different results.)

**Comment by katrinleinweber** [comment]
Following up the `read.table()` discussion (3 :+1:s, 2x caution): Do we need to decide between teaching:

a) either the versatile but multi-faceted foundation here, or
b) the convenient `read_csv()`, which enables us to drop the `stringsAsFactors` part?

2nd Q: How do you all feel about b) plus a new challenge like:

> Importing non-standard data tables with `read.table()`

?

**Comment by Geeketics** [comment]
Hi All,
Popping back into the thread after a long absence. I guess my original point wasn't trying to say "use tidyverse over base R" (or the other way around), but more let learners know that there are multiple ways of doing the same thing, we are teaching them one example way, and to be certain that we all get the same results in the class & can follow along without any surprise differences / errors we'll explicitly write in read.csv() rather than use point & click, which may give different results.

I also think while read_csv() is tidier to look at in that you don't need to explicitly write out `stringsAsFactors = FALSE` the fact that you DO have to write it out with read.csv() gives you a convenient moment to discuss what a factor is & how it can cause you to tear your hair out in frustration if you get hit with a factor column you weren't expecting. I've had some pretty good discussion with classes I've taught about what a factor is and why it is / isn't useful in different situations.

I guess the real question is judging how much your lesson should delve into the tangential possibilities and potential issues vs. sticking on track so as not to overwhelm newcomers with all the different things that could go wrong or be done differently etc.

**Comment by KlausVigo** [comment]
Hi all, 
I agree with the comments of @Geeketics and writing `stringsAsFactors = FALSE` can be an inconvenience. Warning from here on it is getting geeky. It may actually a good exercise to refer and look to the help pages (`?read.csv`,  `?read_csv` and resist googling stackoverflow). We will find lots of arguments are hidden in the `...` argument, there are the description of functions like `read.table` or `read_delim` as well as links to many more. 
Additionally looking at the source code of `read.csv` reveals that 
```
> read.csv
function (file, header = TRUE, sep = ",", quote = "\"", dec = ".", 
    fill = TRUE, comment.char = "", ...) 
read.table(file = file, header = header, sep = sep, quote = quote, 
    dec = dec, fill = fill, comment.char = comment.char, ...)
``` 
`read.csv` is just calling `read.table`, but changes some of the default values. In the exercises and examples for writing functions the focus seems on simplifying a repeated task. It may be a nice exercise to change/extend an existing function. As the source code of functions in R is always at your fingertips they can be (I resist to write here `easily`) modified. Here is an example:   
```
read.csv3 <- function (file, header = TRUE, sep = ",", quote = "\"", dec = ".", 
              fill = TRUE, comment.char = "", stringsAsFactors = FALSE, ...) 
        read.table(file = file, header = header, sep = sep, quote = quote, 
                   dec = dec, fill = fill, comment.char = comment.char,  
                   stringsAsFactors = stringsAsFactors, ...)
``` 
This teaches about default values and the `...` argument and we learn how some of the R masters have written their functions. And this variant of `read.csv` should make most people happy ;)

Unrelated it maybe to consider that some participants may have to install the tidyverse during the workshop and this will take some time to compile (especially `stringr`) and some bandwidth.

**Comment by murraycadzow** [comment]
I'd just like to point out that at least from Rstudio v1.1.442 (maybe earlier versions too) the options from the import dataset menu now specifically gives the option of "From text (base) ..." or "From text (readr) ..."

**Comment by katrinleinweber** [comment]
About tidyverse: there might not be a real need to adapt this lesson, because of https://github.com/UoMResearchIT/r-tidyverse-intro. @diyadas and me agree that we should definitely not have both variants in one lesson at the same time.

**Comment by katrinleinweber** [comment]
Also, there is https://swcarpentry.github.io/r-novice-gapminder/ which is focussing on the tidyverse way of doing things ;-)
